# Limpieza de datos

Vamos a aplicar el [checklist del Banco Mundial](https://dimewiki.worldbank.org/wiki/Checklist:_Data_Cleaning) en Python usando el [SciPy stack](https://www.scipy.org/stackspec.html), principalmente pandas

## 0. Prerequisitos

Aclaración: La próxima celda es para compatilibidad con Colab, NO ES RECOMENDADO realizar pip install desde un notebook.

In [1]:
import sys
in_colab = 'google.colab' in sys.modules

if in_colab:
  BASE_DIR = "https://github.com/DiploDatos/AnalisisYCuracion/raw/master/"
else:
  BASE_DIR = ".."
if 'ftfy' not in sys.modules:
    !pip install 'ftfy<5.6'

El sistema no puede encontrar el archivo especificado.


# 1. Importando los datos

# 1.1. Verificar que no hay problemas en la importación

In [4]:
import pandas as pd
pd.options.display.float_format = '{:.2f}'.format

Veamos de importar datos de proyectos de Kickstarter la plataforma de Crowdsourcing

In [5]:
kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv")


UnicodeDecodeError: 'utf-8' codec can't decode byte 0x99 in position 11: invalid start byte

Por defecto Pandas falla si hay errores para leer datos https://pandas.pydata.org/pandas-docs/stable/io.html#error-handling

Por ahora cambiamos a un archivo más actualizado, volveremos a este error más adelante

In [6]:
kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv")

Veamos los datos cargados en el dataframe

In [7]:
kickstarter_2018

ID                                               name  \
0       1000002330                    The Songs of Adelaide & Abullah   
1       1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
2       1000004038                                     Where is Hank?   
3       1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
4       1000011046  Community Film Project: The Art of Neighborhoo...   
5       1000014025                               Monarch Espresso Bar   
6       1000023410  Support Solar Roasted Coffee & Green Energy!  ...   
7       1000030581  Chaser Strips. Our Strips make Shots their B*tch!   
8       1000034518  SPIN - Premium Retractable In-Ear Headphones w...   
9        100004195  STUDIO IN THE SKY - A Documentary Feature Film...   
10       100004721                                Of Jesus and Madmen   
11       100005484                                   Lisa Lim New CD!   
12      1000055792                                 The Cottage Market   
13      1000056157  G-Spot Place for Gamers to connect with eachot...   
14      1000057089  Tombstone: Old West tabletop game and miniatur...   
15      1000064368                                     Survival Rings   
16      1000064918                                          The Beard   
17      1000068480                   Notes From London: Above & Below   
18      1000070642                Mike Corey's Darkness & Light Album   
19      1000071625                                           Boco Tea   
20      1000072011              CMUK. Shoes: Take on Life Feet First.   
21      1000081649                   MikeyJ clothing brand fundraiser   
22      1000082254                     Alice in Wonderland in G Minor   
23      1000087442  Mountain brew: A quest for alcohol sustainability   
24      1000091520                        The Book Zoo - A Mini-Comic   
25      1000102741  Matt Cavenaugh & Jenny Powers make their 1st a...   
26      1000103948                               Superhero Teddy Bear   
27      1000104688                                Permaculture Skills   
28      1000104953  Rebel Army Origins: The Heroic Story Of Major ...   
29       100011318                      My Moon - Animated Short Film   
...            ...                                                ...   
378631   999832803  MONSTROCARDS- A Creative Drawing Game of Horri...   
378632    99984851                                           "Loreto"   
378633   999865831  STATE OF G - Men's Blazer Jacket & Women's Car...   
378634   999867839                                      Transcaucasia   
378635    99987261  You're Not Cis, You Just Think You Are: My Enb...   
378636   999877902                      Little Saturday's Debut Album   
378637   999878813                           Ways to Look at the Moon   
378638   999880710                                Terminal Uniqueness   
378639   999884445  Superette Tea - Boldly Caffeinated Loose Leaf Tea   
378640   999893629                          Ciridian Helix (Canceled)   
378641    99991751   Here is the oasis for the youngest in the world.   
378642   999929142  ÉPOUVANTAILS : 28mm Figurines de jeux pour KIN...   
378643   999931480                     Google Glass Shield (Canceled)   
378644   999934908                               The Manual Bar Blade   
378645   999941823        The Chronicles of Erinoth - A Graphic Novel   
378646   999943841                                The Dog Coffee Book   
378647   999950812                              The Cease is Increase   
378648   999955533        Cards and Chit Network Season 2! (Canceled)   
378649   999963618      Colorado International Activism Film Festival   
378650    99996661                         Date Seat - Boston iOS App   
378651   999969812                                       AT THE BEACH   
378652   999971898                        Beach Wrestling Documentary   
378653   999972264                                            Islanda   
3

Por defecto solo vemos los valores al comienzo o al final del archivo.

Tomemos una muestra al azar para ver valores más dispersos

In [8]:
# set seed for reproducibility
import numpy as np
np.random.seed(0)

In [9]:
kickstarter_2018.sample(5)

ID                                               name  \
338862   796196901                                 10G Christmas Tree   
277871   483825010                                              Gliff   
47000    123916947                                 STUFFED Food Truck   
111338  1565733636  NeoExodus Adventure: Origin of Man for Pathfin...   
53743   1273544891        NAPOLEON IN NEW YORK! an original TV Series   

               category main_category currency    deadline     goal  \
338862              Art           Art      USD  2010-12-26 10526.00   
277871  Gaming Hardware         Games      USD  2016-03-28 10000.00   
47000       Food Trucks          Food      USD  2015-01-06 60000.00   
111338   Tabletop Games         Games      USD  2012-05-01   500.00   
53743            Comedy  Film & Video      USD  2016-07-26 25000.00   

                   launched  pledged       state  backers country  \
338862  2010-12-08 08:44:04     0.00      failed        0      US   
277871  2016-01-28 04:56:18    51.00      failed        5      US   
47000   2014-11-07 02:24:36    25.00      failed        1      US   
111338  2012-03-15 01:16:10   585.00  successful       17      US   
53743   2016-05-27 00:07:25    25.00      failed        1      US   

        usd pledged  usd_pledged_real  usd_goal_real  
338862         0.00              0.00       10526.00  
277871        51.00             51.00       10000.00  
47000         25.00             25.00       60000.00  
111338       585.00            585.00         500.00  
53743         25.00             25.00       25000.00

No se observa a simple vista ningún problema obvio.

Notar que todos vimos los mismos resultados. Al fijar la semilla no hubo tal azar, esto es algo necesario cuando queremos "reproducir valores aleatorios"

Veamos la descripción del dataset si se corresponde con lo levantado https://www.kaggle.com/kemical/kickstarter-projects/data

In [10]:
pd.DataFrame([["ID", "No description provided", "Numeric"],
["name", "No description provided", "String"],
["category", "No description provided", "String"],
["main_category", "No description provided", "String"],
["currency", "No description provided", "String"],
["deadline", "No description provided", "DateTime"],
["goal", "Goal amount in project currency", "Numeric"],
["launched", "No description provided", "DateTime"],
["pledged", "Pledged amount in the project currency", "Numeric"],
["state", "No description provided", "String"],
["backers", "No description provided", "Numeric"],
["country", "No description provided", "String"],
["usd pledged", "Pledged amount in USD (conversion made by KS)", "Numeric"],
["usd_pledged_real", "Pledged amount in USD (conversion made by fixer.io api)", "Numeric"],
["usd_goal_real", "Goal amount in USD", "Numeric"]], columns=["Field name","Field description", "Type"])

Field name                                  Field description  \
0                 ID                            No description provided   
1               name                            No description provided   
2           category                            No description provided   
3      main_category                            No description provided   
4           currency                            No description provided   
5           deadline                            No description provided   
6               goal                    Goal amount in project currency   
7           launched                            No description provided   
8            pledged             Pledged amount in the project currency   
9              state                            No description provided   
10           backers                            No description provided   
11           country                            No description provided   
12       usd pledged      Pledged amount in USD (conversion made by KS)   
13  usd_pledged_real  Pledged amount in USD (conversion made by fixe...   
14     usd_goal_real                                 Goal amount in USD   

        Type  
0    Numeric  
1     String  
2     String  
3     String  
4     String  
5   DateTime  
6    Numeric  
7   DateTime  
8    Numeric  
9     String  
10   Numeric  
11    String  
12   Numeric  
13   Numeric  
14   Numeric

Ahora veamos los tipos de datos que detectó pandas

In [11]:
kickstarter_2018.dtypes

ID                    int64
name                 object
category             object
main_category        object
currency             object
deadline             object
goal                float64
launched             object
pledged             float64
state                object
backers               int64
country              object
usd pledged         float64
usd_pledged_real    float64
usd_goal_real       float64
dtype: object

Los campos object generalmente son String, entonces parece que no reconoció como fechas en **deadline** y **launched** :(

Veamos los datos un resumen de los datos

In [12]:
kickstarter_2018.describe()

ID         goal     pledged   backers  usd pledged  \
count     378661.00    378661.00   378661.00 378661.00    374864.00   
mean  1074731191.99     49080.79     9682.98    105.62      7036.73   
std    619086204.32   1183391.26    95636.01    907.19     78639.75   
min         5971.00         0.01        0.00      0.00         0.00   
25%    538263516.00      2000.00       30.00      2.00        16.98   
50%   1075275634.00      5200.00      620.00     12.00       394.72   
75%   1610148624.00     16000.00     4076.00     56.00      3034.09   
max   2147476221.00 100000000.00 20338986.27 219382.00  20338986.27   

       usd_pledged_real  usd_goal_real  
count         378661.00      378661.00  
mean            9058.92       45454.40  
std            90973.34     1152950.06  
min                0.00           0.01  
25%               31.00        2000.00  
50%              624.33        5500.00  
75%             4050.00       15500.00  
max         20338986.27   166361390.71

Por defecto se ven los datos numéricos, veamos el resto.

In [13]:
kickstarter_2018.describe(include=['object'])

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   

          deadline             launched   state country  
count       378661               378661  378661  378661  
unique        3164               378089       6      23  
top     2014-08-08  1970-01-01 01:00:00  failed      US  
freq           705                    7  197719  292627

Operemos un cacho sobre los datos de lanzamiento

In [14]:
kickstarter_2018['launched'].min()

'1970-01-01 01:00:00'

Parece funcionar, pero ahora calculemos el rango de fechas de los proyectos

In [15]:
kickstarter_2018['launched'].max() - kickstarter_2018['launched'].min()

TypeError: unsupported operand type(s) for -: 'str' and 'str'

Indiquemos que columnas son fechas como indica la [documentación](https://pandas.pydata.org/pandas-docs/stable/io.html#datetime-handling)

In [45]:
kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"])
kickstarter_2018.dtypes

ID                           int64
name                        object
category                    object
main_category               object
currency                    object
deadline            datetime64[ns]
goal                       float64
launched            datetime64[ns]
pledged                    float64
state                       object
backers                      int64
country                     object
usd pledged                float64
usd_pledged_real           float64
usd_goal_real              float64
dtype: object

Ahora vemos que esas columnas fueron reconocidas como fechas

Veamos la misma muestra de nuevo

In [46]:
kickstarter_2018.sample(5)

ID                                     name     category  \
299667   595962034                      Button Poetry Live!  Mixed Media   
181674  1924707671         C STREET 2012 : Tbilisi, Georgia  World Music   
137583  1698707842                    Dérive's Next Project         Punk   
296861   581269566  Photo Book - World Santa Claus Congress   Photobooks   
66362   1337585114      Kickstart CLE Brewing to greatness!       Drinks   

       main_category currency   deadline      goal            launched  \
299667           Art      USD 2015-09-18  10000.00 2015-08-19 19:34:20   
181674         Music      USD 2012-06-07   5000.00 2012-05-08 18:22:59   
137583         Music      USD 2014-07-06   1200.00 2014-06-08 17:58:37   
296861   Photography      DKK 2017-04-14 110000.00 2017-03-14 23:45:35   
66362           Food      USD 2017-08-19   6500.00 2017-07-20 21:22:43   

        pledged       state  backers country  usd pledged  usd_pledged_real  \
299667 18216.27  successful      455      US     18216.27          18216.27   
181674  7210.69  successful       82      US      7210.69           7210.69   
137583  1255.66  successful       33      US      1255.66           1255.66   
296861   462.00      failed        5      DK         0.00             66.46   
66362    250.00      failed        5      US        75.00            250.00   

        usd_goal_real  
299667       10000.00  
181674        5000.00  
137583        1200.00  
296861       15823.47  
66362         6500.00

Y veamos el resumen de los datos

In [47]:
kickstarter_2018.describe(include='all')

ID                      name        category main_category  \
count      378661.00                    378657          378661        378661   
unique           nan                    375764             159            15   
top              nan  New EP/Music Development  Product Design  Film & Video   
freq             nan                        41           22314         63585   
first            nan                       NaN             NaN           NaN   
last             nan                       NaN             NaN           NaN   
mean   1074731191.99                       NaN             NaN           NaN   
std     619086204.32                       NaN             NaN           NaN   
min          5971.00                       NaN             NaN           NaN   
25%     538263516.00                       NaN             NaN           NaN   
50%    1075275634.00                       NaN             NaN           NaN   
75%    1610148624.00                       NaN             NaN           NaN   
max    2147476221.00                       NaN             NaN           NaN   

       currency             deadline         goal             launched  \
count    378661               378661    378661.00               378661   
unique       14                 3164          nan               378089   
top         USD  2014-08-08 00:00:00          nan  1970-01-01 01:00:00   
freq     295365                  705          nan                    7   
first       NaN  2009-05-03 00:00:00          nan  1970-01-01 01:00:00   
last        NaN  2018-03-03 00:00:00          nan  2018-01-02 15:02:31   
mean        NaN                  NaN     49080.79                  NaN   
std         NaN                  NaN   1183391.26                  NaN   
min         NaN                  NaN         0.01                  NaN   
25%         NaN                  NaN      2000.00                  NaN   
50%         NaN                  NaN      5200.00                  NaN   
75%         NaN                  NaN     16000.00                  NaN   
max         NaN                  NaN 100000000.00                  NaN   

           pledged   state   backers country  usd pledged  usd_pledged_real  \
count    378661.00  378661 378661.00  378661    374864.00         378661.00   
unique         nan       6       nan      23          nan               nan   
top            nan  failed       nan      US          nan               nan   
freq           nan  197719       nan  292627          nan               nan   
first          nan     NaN       nan     NaN          nan               nan   
last           nan     NaN       nan     NaN          nan               nan   
mean       9682.98     NaN    105.62     NaN      7036.73           9058.92   
std       95636.01     NaN    907.19     NaN     78639.75          90973.34   
min           0.00     NaN      0.00     NaN         0.00              0.00   
25%          30.00     NaN      2.00     NaN        16.98             31.00   
50%         620.00     NaN     12.00     NaN       394.72            624.33   
75%        4076.00     NaN     56.00     NaN      3034.09           4050.00   
max    20338986.27     NaN 219382.00     NaN  20338986.27       20338986.27   

        usd_goal_real  
count       378661.00  
unique            nan  
top               nan  
freq              nan  
first             nan  
last              nan  
mean         45454.40  
std        1152950.06  
min              0.01  
25%           2000.00  
50%           5500.00  
75%          15500.00  
max      166361390.71

Podemos ver que tenemos primero y último en el resumen de las columnas de fechas.

Ahora deberíamos poder calcular el rango de fechas de lanzamietos

In [48]:
kickstarter_2018['launched'].max() - kickstarter_2018['launched'].min()

Timedelta('17533 days 14:02:31')

# 1.2. Asegurar de tener ids/claves únicas

Chequear que no hay datos duplicados

In [49]:
kickstarter_2018.shape

(378661, 15)

Pandas soporta índices en los DataFrames vamos a recargar el conjunto de datos

In [113]:
kickstarter_2018 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv", 
                               parse_dates=["deadline","launched"],
                               index_col=['ID'])

In [51]:
kickstarter_2018

name  \
ID                                                              
1000002330                    The Songs of Adelaide & Abullah   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET   
1000004038                                     Where is Hank?   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
1000011046  Community Film Project: The Art of Neighborhoo...   
1000014025                               Monarch Espresso Bar   
1000023410  Support Solar Roasted Coffee & Green Energy!  ...   
1000030581  Chaser Strips. Our Strips make Shots their B*tch!   
1000034518  SPIN - Premium Retractable In-Ear Headphones w...   
100004195   STUDIO IN THE SKY - A Documentary Feature Film...   
100004721                                 Of Jesus and Madmen   
100005484                                    Lisa Lim New CD!   
1000055792                                 The Cottage Market   
1000056157  G-Spot Place for Gamers to connect with eachot...   
1000057089  Tombstone: Old West tabletop game and miniatur...   
1000064368                                     Survival Rings   
1000064918                                          The Beard   
1000068480                   Notes From London: Above & Below   
1000070642                Mike Corey's Darkness & Light Album   
1000071625                                           Boco Tea   
1000072011              CMUK. Shoes: Take on Life Feet First.   
1000081649                   MikeyJ clothing brand fundraiser   
1000082254                     Alice in Wonderland in G Minor   
1000087442  Mountain brew: A quest for alcohol sustainability   
1000091520                        The Book Zoo - A Mini-Comic   
1000102741  Matt Cavenaugh & Jenny Powers make their 1st a...   
1000103948                               Superhero Teddy Bear   
1000104688                                Permaculture Skills   
1000104953  Rebel Army Origins: The Heroic Story Of Major ...   
100011318                       My Moon - Animated Short Film   
...                                                       ...   
999832803   MONSTROCARDS- A Creative Drawing Game of Horri...   
99984851                                             "Loreto"   
999865831   STATE OF G - Men's Blazer Jacket & Women's Car...   
999867839                                       Transcaucasia   
99987261    You're Not Cis, You Just Think You Are: My Enb...   
999877902                       Little Saturday's Debut Album   
999878813                            Ways to Look at the Moon   
999880710                                 Terminal Uniqueness   
999884445   Superette Tea - Boldly Caffeinated Loose Leaf Tea   
999893629                           Ciridian Helix (Canceled)   
99991751     Here is the oasis for the youngest in the world.   
999929142   ÉPOUVANTAILS : 28mm Figurines de jeux pour KIN...   
999931480                      Google Glass Shield (Canceled)   
999934908                                The Manual Bar Blade   
999941823         The Chronicles of Erinoth - A Graphic Novel   
999943841                                 The Dog Coffee Book   
999950812                               The Cease is Increase   
999955533         Cards and Chit Network Season 2! (Canceled)   
999963618       Colorado International Activism Film Festival   
99996661                           Date Seat - Boston iOS App   
999969812                                        AT THE BEACH   
999971898                         Beach Wrestling Documentary   
999972264                                             Islanda   
999975836               Homemade fresh dog food, Cleveland OH   
999976312                          Angela's Poetry (Canceled)   
999976400   ChknTruk Nationwide Charity Drive 2014 (Canceled)   
999977640                                           The Tribe   
999986353   Walls of Remedy- New lesbian Romantic Comedy f...   
999987933                            BioDefense Education Kit   
999988282                  

In [52]:
kickstarter_2018.shape

(378661, 14)

De esta forma podemos buscar por el índice

In [53]:
kickstarter_2018.loc[999988282]

name                Nou Renmen Ayiti!  We Love Haiti!
category                              Performance Art
main_category                                     Art
currency                                          USD
deadline                          2011-08-16 00:00:00
goal                                          2000.00
launched                          2011-07-19 09:07:47
pledged                                        524.00
state                                          failed
backers                                            17
country                                            US
usd pledged                                    524.00
usd_pledged_real                               524.00
usd_goal_real                                 2000.00
Name: 999988282, dtype: object

También podemos verificar si hay filas de contenidos duplicado

In [54]:
kickstarter_2018[kickstarter_2018.duplicated()]

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

Como Pandas acepta valores duplicados en los índices también debemos verificar ahí

In [55]:
pd.Series(kickstarter_2018.index,dtype=str).describe()

count        378661
unique       378661
top       439416501
freq              1
Name: ID, dtype: object

In [17]:
kickstarter_2018[kickstarter_2018.index.duplicated()]

Empty DataFrame
Columns: [ID, name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd pledged, usd_pledged_real, usd_goal_real]
Index: []

Repasamos con un ejemplo de juguete

In [18]:
csv='1,2\n3,3\n1,3'
print(csv)

1,2
3,3
1,3


In [19]:
from io import StringIO
df = pd.read_csv(StringIO(csv), names=['id','value'], index_col='id')
df

value
id       
1       2
3       3
1       3

In [20]:
df[df.duplicated()]

value
id       
1       3

In [21]:
df[df.index.duplicated(keep=False)]

value
id       
1       2
1       3

#### Ejercicio 1:

Armar una tabla con todos los proyectos con nombres duplicados, ordenados para revisar agrupados. 

In [65]:
repetidos = kickstarter_2018[kickstarter_2018.name.duplicated(keep=False)].sort_values(['name'], ascending=[True]) 

repetidos.head(5)

name     category  \
ID                                                                           
816998285   "...The Last shall be first..." LODB Lifestyle...      Fashion   
815783250   "...The Last shall be first..." LODB Lifestyle...      Fashion   
1010584633                                    "A Fresh Start"       Shorts   
713417995                                     "A Fresh Start"  Documentary   
1880084695      "American Sports Stories" - An Athletic Quest       Shorts   

           main_category currency   deadline      goal            launched  \
ID                                                                           
816998285        Fashion      USD 2014-02-09   6500.00 2013-12-19 21:51:50   
815783250        Fashion      USD 2014-03-01   6500.00 2014-01-03 20:39:16   
1010584633  Film & Video      USD 2011-09-25   3000.00 2011-08-28 21:27:52   
713417995   Film & Video      USD 2013-03-27   5000.00 2013-01-26 03:51:47   
1880084695  Film & Video      USD 2015-08-26 100000.00 2015-06-27 02:02:00   

            pledged       state  backers country  usd pledged  \
ID                                                              
816998285      0.00    canceled        0      US         0.00   
815783250     37.00      failed        3      US        37.00   
1010584633  3000.00  successful       24      US      3000.00   
713417995   1417.00      failed       26      US      1417.00   
1880084695   100.00      failed        1      US       100.00   

            usd_pledged_real  usd_goal_real  
ID                                           
816998285               0.00        6500.00  
815783250              37.00        6500.00  
1010584633           3000.00        3000.00  
713417995            1417.00        5000.00  
1880084695            100.00      100000.00

5125 proyectos poseen nombres que estan repetidos

# 1.3. Despersonalizar datos y guardarlos en un nuevo archivo

Hay muuuchas técnicas para despersonalizar datos.

Para ilustrar mostramos las ofrecidas por Google https://cloud.google.com/dlp/docs/transformations-reference:

* **Reemplazo**: Reemplaza cada valor de entrada con un valor determinado.
* **Ocultamiento**: Quita un valor y lo oculta.
* **Enmascaramiento con caracteres**: Enmascara una string por completo o parcialmente mediante el reemplazo de un número determinado de caracteres con un carácter fijo especificado..
* **Seudonimización mediante el reemplazo de un valor de entrada con un hash criptográfico**: Reemplaza valores de entrada con una string hexadecimal de 32 bytes mediante una clave de encriptación de datos.
* **Cambio de fechas**: Cambia las fechas por un número de días al azar, con la opción de ser coherente en el mismo contexto..
* **Seudonimización mediante el reemplazo con token de preservación de formato criptográfico**: Reemplaza un valor de entrada con un token, o valor sustituto, de la misma longitud mediante la encriptación de preservación de formato (FPE) con el modo de operación FFX. Esto permite que se use el resultado en sistemas con validación de formato o que necesitan aparecer como reales a pesar de que la información no se revela.
* **Valores de depósito con base en rangos de tamaño fijos**: Enmascara los valores de entrada y los reemplaza por depósitos, o rangos dentro de los cuales se encuentra el valor de entrada.
* **Valores de depósito con base en rangos de tamaño personalizados**: Valores de entrada de depósito con base en rangos configurables por el usuario y valores de reemplazo.
* **Extracción de datos de tiempo**: EExtrae o preserva una porción de los valores Date, Timestamp y TimeOfDay.

In [66]:
from hashlib import md5

In [71]:
kickstarter_2018['name'].apply(md5)

TypeError: Unicode-objects must be encoded before hashing

In [73]:
def hashit(val):
    return md5(val.encode('utf-8'))

kickstarter_2018['name'].apply(hashit)

AttributeError: 'float' object has no attribute 'encode'

In [74]:
def hashit(val):
    try:
        return md5(val.encode('utf-8'))
    except Exception as e:
        print(val, type(val))
        raise(e)

kickstarter_2018['name'].apply(hashit)

nan <class 'float'>


AttributeError: 'float' object has no attribute 'encode'

In [78]:
def hashit(val):
    if isinstance(val, float): 
        return str(val)
    return md5(val.encode('utf-8')).hexdigest()


kickstarter_2018['name'].apply(hashit)

ID
1000002330    a6828ae8a2eca25f0dd7035efc0af0a0
1000003930    81609b3bdc0b96f429672d69702f2524
1000004038    c12f5c3bace2f0213cdb2679a265dca0
1000007540    4dbdcf09c86bbf5683ec39bc57b77f81
1000011046    9c01404a2ef702811c2088ce139042ad
1000014025    3a26b565c8396d4867dda27816836e04
1000023410    1602dce9289c950176c41ce124944335
1000030581    1abb0a27e1a3ce5a79965e3022c6bddc
1000034518    3b3a87aebec0d9c1b3a4a5e4da7a88a8
100004195     d7c53ef574a108d857d52087bb685244
100004721     eba20b66506f511bfacfe843176f5e15
100005484     9df25d0c0e2f8ea9de303f3d4efaf3c8
1000055792    717f62eae71408f00fa49227da83feae
1000056157    6532f0fa53550fd516921a2d7a1121d6
1000057089    799a510f874b6d47c1fde736d22cae3e
1000064368    93ebbffc751f9cc8d31e0147b7b1c7e3
1000064918    c10592831d537538905aefc12d54b49d
1000068480    0f109f9875645b801843775bbeadb1f1
1000070642    2d3d4b4267af87948bf7ac6d3ca13b54
1000071625    26c10e7de5df6269e71df1bf49ff4da9
1000072011    459cec5845add6a7f8fb1498cfc0e618
1000081649

#### Ejercicio 2:

Verificar que los proyectos que tienen nombres duplicados también tienen el hash de nombre duplicado

In [83]:
repetidos['name'].apply(hashit)

ID
816998285     0c7a251ffe4c7834cbc4b04906952ff1
815783250     0c7a251ffe4c7834cbc4b04906952ff1
1010584633    67554ab4203d95f2f2f05365f768206e
713417995     67554ab4203d95f2f2f05365f768206e
1880084695    19a82bc4c5961834282575d07d9b5f7c
422509694     19a82bc4c5961834282575d07d9b5f7c
866225086     b4ca2044f21f67e13ddb7583f59189a7
1760892298    b4ca2044f21f67e13ddb7583f59189a7
1999958521    bed6c6cbc0540c8d40b7e2311806cc5c
870322805     bed6c6cbc0540c8d40b7e2311806cc5c
1413519345    24873b2974fe8222256c6cc800a2b361
1637334549    24873b2974fe8222256c6cc800a2b361
1656736114    5f38010cf3ee47694a209c9a8403acfa
159049492     5f38010cf3ee47694a209c9a8403acfa
72198224      b2226ed5c4031a0c849ac470e9910f7d
996180421     b2226ed5c4031a0c849ac470e9910f7d
306461885     2587a1a381b7ff6d58e564532406b035
1023301684    2587a1a381b7ff6d58e564532406b035
516801004     0e9a80859587c8f75672828db5fb9118
2007967203    0e9a80859587c8f75672828db5fb9118
1223071711    2727d1da48d26f0f96b6e1d6e591c447
2136175444

Se verifica que se repiten dado que estan ordenados

# 1.4. Nunca modificar los datos crudos u originales


In [84]:
if not in_colab:
  kickstarter_2018.to_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv")

FileNotFoundError: [Errno 2] No such file or directory: '../output/ks-projects-201801-for-pandas.csv'


# 2. Pasos necesarios


In [85]:
#import pandas as pd
if not in_colab:
  kickstarter_2018 = pd.read_csv(BASE_DIR + "/output/ks-projects-201801-for-pandas.csv",
                        index_col='ID',
                        parse_dates=['deadline','launched'])

FileNotFoundError: [Errno 2] File b'../output/ks-projects-201801-for-pandas.csv' does not exist: b'../output/ks-projects-201801-for-pandas.csv'

In [86]:
kickstarter_2018.describe(include='all')

name        category main_category currency  \
count                     378657          378661        378661   378661   
unique                    375764             159            15       14   
top     New EP/Music Development  Product Design  Film & Video      USD   
freq                          41           22314         63585   295365   
first                        NaN             NaN           NaN      NaN   
last                         NaN             NaN           NaN      NaN   
mean                         NaN             NaN           NaN      NaN   
std                          NaN             NaN           NaN      NaN   
min                          NaN             NaN           NaN      NaN   
25%                          NaN             NaN           NaN      NaN   
50%                          NaN             NaN           NaN      NaN   
75%                          NaN             NaN           NaN      NaN   
max                          NaN             NaN           NaN      NaN   

                   deadline         goal             launched     pledged  \
count                378661    378661.00               378661   378661.00   
unique                 3164          nan               378089         nan   
top     2014-08-08 00:00:00          nan  1970-01-01 01:00:00         nan   
freq                    705          nan                    7         nan   
first   2009-05-03 00:00:00          nan  1970-01-01 01:00:00         nan   
last    2018-03-03 00:00:00          nan  2018-01-02 15:02:31         nan   
mean                    NaN     49080.79                  NaN     9682.98   
std                     NaN   1183391.26                  NaN    95636.01   
min                     NaN         0.01                  NaN        0.00   
25%                     NaN      2000.00                  NaN       30.00   
50%                     NaN      5200.00                  NaN      620.00   
75%                     NaN     16000.00                  NaN     4076.00   
max                     NaN 100000000.00                  NaN 20338986.27   

         state   backers country  usd pledged  usd_pledged_real  usd_goal_real  
count   378661 378661.00  378661    374864.00         378661.00      378661.00  
unique       6       nan      23          nan               nan            nan  
top     failed       nan      US          nan               nan            nan  
freq    197719       nan  292627          nan               nan            nan  
first      NaN       nan     NaN          nan               nan            nan  
last       NaN       nan     NaN          nan               nan            nan  
mean       NaN    105.62     NaN      7036.73           9058.92       45454.40  
std        NaN    907.19     NaN     78639.75          90973.34     1152950.06  
min        NaN      0.00     NaN         0.00              0.00           0.01  
25%        NaN      2.00     NaN        16.98             31.00        2000.00  
50%        NaN     12.00     NaN       394.72            624.33        5500.00  
75%        NaN     56.00     NaN      3034.09           4050.00       15500.00  
max        NaN 219382.00     NaN  20338986.27       20338986.27   166361390.71

## 2.1. Etiquetas de variables/columnas: no usar caracteres especiales



In [87]:
# helpful character encoding module
import chardet

¿Por qué? Por que aun hay limitaciones para trabajar con estos caracteres.

¿Cúales son los caracteres "normales"? 

In [89]:
import string

string.ascii_letters + string.digits

'abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789'

¿Qué es ASCII? Es una de las primeras codificaciones (o encoding) disponibles. Ver https://en.wikipedia.org/wiki/ASCII

Un encoding es un mapa de caracteres a una representación en bits (por ejemplo 1000001). ASCII es uno de los primeros estandares para interoperatividad entre computadoras (antes cada fabricante de computadora usa su propia codificación), contempla sólo los caracteres ingleses y usa 7 bits, por ejemplo *1000001* codifica el caracter *A*

Cómo sólo representaban los caracteres ingleses, empiezan a aparecer variantes para distintos idiomas, Microsoft hizo los propios para internacionalizar Windows, hubo otro estandar ISO 8859, hasta que llegó el éstandar Unicode que hizo múltiples encodings pero cada uno contempla múltiples idiomas e incluso es extensible a futuro. UTF-8 es de esos encodings el más utilizado. Ver https://en.wikipedia.org/wiki/Unicode

En particular, Python 3 utiliza UTF-8 por defecto (lo que no sucedía en Python 2).

Veamos un ejemplo práctico

In [91]:
# start with a string
before = "This is the euro symbol: €"

# check to see what datatype it is
type(before)

str

In [94]:
# encode it to a different encoding, replacing characters that raise errors
after = before.encode("utf-8", errors = "replace")

# check the type
type(after)

bytes

In [31]:
# take a look at what the bytes look like
after

b'This is the euro symbol: \xe2\x82\xac'

In [96]:
# convert it back to utf-8
print(after.decode("utf-8"))

This is the euro symbol: €


In [101]:
# try to decode our bytes with the ascii encoding
print(after.decode("ascii"))

This is the euro symbol: ?


In [114]:
# start with a string
before = "This is the euro symbol: €"

# encode it to a different encoding, replacing characters that raise errors
after = before.encode("ascii", errors = "replace")

# convert it back to utf-8
print(after.decode("ascii"))

# We've lost the original underlying byte string! It's been 
# replaced with the underlying byte string for the unknown character :(

This is the euro symbol: ?


Este error ya lo vimos...

In [115]:
kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv")

UnicodeDecodeError: 'utf-8' codec can't decode byte 0x99 in position 11: invalid start byte

El error `UnicodeDecodeError` nos indica que el archivo no estaría en Unicode. Si no está en algún metadato o no fue indicado es dificil detectar el encoding, por suerte hay un paquete para esto en Python, que aplica unas heurísticas para detectar encoding.

In [116]:
# In Colab we need to download data locally
if in_colab:
  !wget -c $BASE_DIR/input/kickstarter-projects/ks-projects-201612.csv
  DOWN_DIR = "./"
else:
  DOWN_DIR = BASE_DIR + "/input/kickstarter-projects/"
  
# look at the first ten thousand bytes to guess the character encoding
with open(DOWN_DIR + "/ks-projects-201612.csv", 'rb') as rawdata:
    result = chardet.detect(rawdata.read(10000))

# check what the character encoding might be
result

{'encoding': 'Windows-1252', 'confidence': 0.73, 'language': ''}

Entonces le podemos indicar el encoding al leer el archivos

In [117]:
# read in the file with the encoding detected by chardet
kickstarter_2016 = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201612.csv", encoding='Windows-1252')

# look at the first few lines
kickstarter_2016.head()

ID                                               name   \
0  1000002330                    The Songs of Adelaide & Abullah   
1  1000004038                                     Where is Hank?   
2  1000007540  ToshiCapital Rekordz Needs Help to Complete Album   
3  1000011046  Community Film Project: The Art of Neighborhoo...   
4  1000014025                               Monarch Espresso Bar   

        category  main_category  currency             deadline   goal   \
0          Poetry     Publishing       GBP  2015-10-09 11:36:00   1000   
1  Narrative Film   Film & Video       USD  2013-02-26 00:20:50  45000   
2           Music          Music       USD  2012-04-16 04:24:11   5000   
3    Film & Video   Film & Video       USD  2015-08-29 01:00:00  19500   
4     Restaurants           Food       USD  2016-04-01 13:38:27  50000   

             launched  pledged       state  backers  country  usd pledged   \
0  2015-08-11 12:12:28        0      failed        0       GB            0   
1  2013-01-12 00:20:50      220      failed        3       US          220   
2  2012-03-17 03:24:11        1      failed        1       US            1   
3  2015-07-04 08:35:03     1283    canceled       14       US         1283   
4  2016-02-26 13:38:27    52375  successful      224       US        52375   

  Unnamed: 13 Unnamed: 14 Unnamed: 15  Unnamed: 16  
0         NaN         NaN         NaN          nan  
1         NaN         NaN         NaN          nan  
2         NaN         NaN         NaN          nan  
3         NaN         NaN         NaN          nan  
4         NaN         NaN         NaN          nan

Hay otro problema con los caracteres que es más sutíl: **[Mojibake](https://en.wikipedia.org/wiki/Mojibake)**

Sucede al leer UTF-8 como ASCII extendido (ie. `Windows-*`, `iso-8859-*` entre otros)

Ejemplo de mojibake:

In [118]:
"cigüeña".encode('utf-8').decode('iso-8859-1')

'cigüeña'

Para esto vamos a usar un módulo específico [ftfy](https://ftfy.readthedocs.io/en/latest/). 

Que nos permite ordenar cadenas por rareza

In [119]:
import ftfy.badness as bad

In [120]:
def weird(val):
    if isinstance(val, float): 
        return 0
    return bad.sequence_weirdness(val)


kickstarter_2018['name_weirdness'] = kickstarter_2018['name'].apply(weird)
kickstarter_2018[kickstarter_2018['name_weirdness'] > 1]

name  \
ID                                                              
1221990593             Support PMCD´s ¨From a Different Clay¨   
1234089244  Cracking (Mastering) the Metaphor 
(Political)...   
124795481                                    µGALE (Canceled)   
1468489999  µCurrent GOLD - Precision Multimeter Current A...   
1610503461  µPeek - The Professional Microscope that Fits ...   
16174550                                       PARALELO 19 Nº   
1719215256           Kickstarter Gold: Manual Coffeemaker Nº3   
1837094178  ¨Midgard's taigh-seinnse/ Eröffnung Mittelalte...   
2056123754      ZEIº - the most simple time tracking solution   
2068157547  Ballet Folklorico Infantil de la Universidad ...   
2096648634                          Funky City : a funky rpg   
2102265118      Music Album ¨Dreams¨ Uplifting Tunes with Rap   
237780030   Solderdoodle Pro: Over 700ºF USB Rechargeable ...   
383244441            Virtual Roller Coaster in 4th Dimension   
458626388   Solderdoodle Pro: 700ºF USB Rechargeable Solde...   
466898065                                Manual Tea Maker Nº1   
516833776    µTracker - extremly Power Saving GPS + GSM Board   
549309727   Higher Grounds Coffee Company Mobile Coffee Tr...   
551994337   ?? cybertwee ?? ¨*:·. bake sale on the deep we...   
558638957                                         Mµ Magazine   
718763713      Not Twins - New EP! "The View from Down Here"   
805839986   µRuler - Measure size in the micro-world under...   
900880068                                Newton ¨El Veterano¨   
960506632                              Degreesº to Extinction   
963442008          theNewerYork Lit Mag: We M?§§ w/ Yoµr H£åd   
98169605    µTøøl - The Ultimate pocket tool! Credit card ...   

                   category main_category currency   deadline      goal  \
ID                                                                        
1221990593            Dance         Dance      USD 2011-06-08   1500.00   
1234089244       Nonfiction    Publishing      USD 2013-10-27  30000.00   
124795481         Wearables    Technology      EUR 2015-04-17  18000.00   
1468489999         Hardware    Technology      AUD 2013-12-31   9900.00   
1610503461       Technology    Technology      CHF 2016-01-09 125000.00   
16174550        Photography   Photography      MXN 2017-04-15  15000.00   
1719215256   Product Design        Design      USD 2017-07-22  60000.00   
1837094178      Restaurants          Food      CHF 2016-09-29  50000.00   
2056123754       Technology    Technology      EUR 2016-10-26  75000.00   
2068157547              Art           Art      MXN 2017-04-03  40000.00   
2096648634      Video Games         Games      USD 2017-04-08   5000.00   
2102265118          Hip-Hop         Music      EUR 2017-12-24  28000.00   
237780030   DIY Electronics    Technology      USD 2014-08-19  37800.00   
383244441       Makerspaces    Technology      MXN 2017-02-25 459900.00   
458626388   DIY Electronics    Technology      USD 2014-09-24  11400.00   
466898065    Product Design        Design      USD 2016-08-20  40000.00   
516833776          Hardware    Technology      EUR 2016-12-09  89000.00   
549309727       Food Trucks          Food      USD 2018-01-12  10000.00   
551994337       Digital Art           Art      USD 2015-09-18    100.00   
558638957             Zines    Publishing      CAD 2016-12-02   5000.00   
718763713        Indie Rock         Music      USD 2017-03-27   3500.00   
805839986        Technology    Technology      CAD 2017-04-07   2000.00   
900880068             Music         Music      EUR 2016-12-10   2747.00   
960506632       Documentary  Film & Video      USD 2016-12-21  30000.00   
963442008       Periodicals    Publishing      USD 2012-03-25   7000.00   
98169605         Technology    Technology      NZD 2016-01-01  20000.00   

                      launched   pledged       state  backers country  \
ID                                                            

Volviendo a la consigna original, chequeamos que no haya caracteres fuera de a-Z, 0-9 y _ en los nombres de columnas

In [121]:
kickstarter_2018.columns[~kickstarter_2018.columns.str.match(r'^(\w+)$')]

Index(['usd pledged'], dtype='object')

In [122]:
kickstarter_2018.usd_pledged.head()

AttributeError: 'DataFrame' object has no attribute 'usd_pledged'

In [123]:
kickstarter_2018.columns = kickstarter_2018.columns.str.replace(' ', '_')
kickstarter_2018.head()

name        category  \
ID                                                                              
1000002330                    The Songs of Adelaide & Abullah          Poetry   
1000003930      Greeting From Earth: ZGAC Arts Capsule For ET  Narrative Film   
1000004038                                     Where is Hank?  Narrative Film   
1000007540  ToshiCapital Rekordz Needs Help to Complete Album           Music   
1000011046  Community Film Project: The Art of Neighborhoo...    Film & Video   

           main_category currency   deadline     goal            launched  \
ID                                                                          
1000002330    Publishing      GBP 2015-10-09  1000.00 2015-08-11 12:12:28   
1000003930  Film & Video      USD 2017-11-01 30000.00 2017-09-02 04:43:57   
1000004038  Film & Video      USD 2013-02-26 45000.00 2013-01-12 00:20:50   
1000007540         Music      USD 2012-04-16  5000.00 2012-03-17 03:24:11   
1000011046  Film & Video      USD 2015-08-29 19500.00 2015-07-04 08:35:03   

            pledged     state  backers country  usd_pledged  usd_pledged_real  \
ID                                                                              
1000002330     0.00    failed        0      GB         0.00              0.00   
1000003930  2421.00    failed       15      US       100.00           2421.00   
1000004038   220.00    failed        3      US       220.00            220.00   
1000007540     1.00    failed        1      US         1.00              1.00   
1000011046  1283.00  canceled       14      US      1283.00           1283.00   

            usd_goal_real  name_weirdness  
ID                                         
1000002330        1533.95               0  
1000003930       30000.00               0  
1000004038       45000.00               0  
1000007540        5000.00               0  
1000011046       19500.00               0

In [124]:
kickstarter_2018.usd_pledged.head()

ID
1000002330      0.00
1000003930    100.00
1000004038    220.00
1000007540      1.00
1000011046   1283.00
Name: usd_pledged, dtype: float64

#### Ejercicio 3

Comparar la cantidad de nombres raros en kickstarter_2018 con la que obtenemos al cargar 'ks-projects-201801.csv' con encoding iso-8859-1.

In [141]:
kickstarter_2018_iso = pd.read_csv(BASE_DIR + "/input/kickstarter-projects/ks-projects-201801.csv",
                                parse_dates=["deadline","launched"],
                               index_col=['ID'],encoding='iso-8859-1')

kickstarter_2018_iso['name_weirdness'] = kickstarter_2018_iso['name'].apply(weird)

In [142]:
kickstarter_2018[kickstarter_2018['name_weirdness'] > 1].name.count()

26

In [144]:
kickstarter_2018_iso[kickstarter_2018_iso['name_weirdness'] > 1].name.count()

7682

En el importado normal tenemos 26 valores "raros" mientras que en el importado con el encoding iso-8859-1 tenemos 7682

In [148]:
kickstarter_2018[['name','name_weirdness']][kickstarter_2018['name_weirdness'] > 1].sort_values(['name_weirdness'], ascending=[False]).head(5) 

name  name_weirdness
ID                                                                           
900880068                                Newton ¨El Veterano¨               4
2102265118      Music Album ¨Dreams¨ Uplifting Tunes with Rap               4
1221990593             Support PMCD´s ¨From a Different Clay¨               3
458626388   Solderdoodle Pro: 700ºF USB Rechargeable Solde...               2
963442008          theNewerYork Lit Mag: We M?§§ w/ Yoµr H£åd               2

In [150]:
kickstarter_2018_iso[['name','name_weirdness']][kickstarter_2018_iso['name_weirdness'] > 1].sort_values(['name_weirdness'], ascending=[False]).head(5) 

name  name_weirdness
ID                                                                           
975164714   Michale Graves “Wanderer” Album – Follow...              20
1627300385  1960‘s, ‘70’s & ‘80’s Race Car Photo...              20
507403134   ••• New Music and Video Art ••• CD...              18
1313349951  Own A Maine Lighthouse -- 43° 39’ 21” N, ...              18
592878198   VINCENT, a Thriller -“Dr.Jeckyl/Hyde” meet...              16

Viendo los 5 primeros de cada uno vemos los problemas graves de formato que tuvo el encoding iso-8859-1

## 2.2. Tratar valores faltantes

Veamos cuantos valores nulos tenemos

In [151]:
# get the number of missing data points per column
missing_values_count = kickstarter_2018.isnull().sum()

# look at the # of missing points in the first ten columns
missing_values_count[missing_values_count > 0]

name              4
usd_pledged    3797
dtype: int64

Esta medida en sí, no dice nada más que con cuantos valores debemos lidiar. 

Primero tenemos que ver si en términos relativos su impacto.

In [155]:
len(kickstarter_2018.dropna())/len(kickstarter_2018)

0.9899619976707398

In [161]:
len(kickstarter_2018.dropna(subset=['name']))/len(kickstarter_2018)

0.9999894364616372

También debemos ver que significan en términos de representatividad e importancia de nuestro análisis.

In [163]:
kickstarter_2018[kickstarter_2018.name.isnull()]

name        category main_category currency   deadline      goal  \
ID                                                                            
1848699072  NaN  Narrative Film  Film & Video      USD 2012-02-29 200000.00   
634871725   NaN     Video Games         Games      GBP 2013-01-06   2000.00   
648853978   NaN  Product Design        Design      USD 2016-07-18   2500.00   
796533179   NaN        Painting           Art      USD 2011-12-05  35000.00   

                      launched  pledged      state  backers country  \
ID                                                                    
1848699072 2012-01-01 12:35:31   100.00     failed        1      US   
634871725  2012-12-19 23:57:48   196.00     failed       12      GB   
648853978  2016-06-18 05:01:47     0.00  suspended        0      US   
796533179  2011-11-06 23:55:55   220.00     failed        5      US   

            usd_pledged  usd_pledged_real  usd_goal_real  name_weirdness  
ID                                                                        
1848699072       100.00            100.00      200000.00               0  
634871725        317.73            316.05        3224.97               0  
648853978          0.00              0.00        2500.00               0  
796533179        220.00            220.00       35000.00               0

Ejemplo, cuantos de los proyectos suspendidos no tienen nombre

In [164]:
len(kickstarter_2018[kickstarter_2018.state == 'suspended'].dropna(subset=['name'])) / \
    len(kickstarter_2018[kickstarter_2018.state == 'suspended'])

0.9994582881906826

En este caso, creemos que es seguro descartar los proyectos sin nombre

In [165]:
kickstarter_2018 = kickstarter_2018.dropna(subset=['name'])

In [166]:
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

usd_pledged    3797
dtype: int64

In [167]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()]

name      category  \
ID                                                                            
1000694855                      STREETFIGHTERZ WHEELIE MURICA  Film & Video   
100149523                         Duncan Woods - Chameleon EP         Music   
1003023003          The Making of Ashley Kelley's Debut Album         Music   
1003130892                       Butter Side Down Debut Album         Music   
1003629045                            Chase Goehring debut EP         Music   
1004013077           Spencer Capier Instrumental Project 2015         Music   
1004126342  LUKAS LIGETI'S 50TH BIRTHDAY FESTIVAL: ORIGINA...         Music   
1004255433      The Battle For Breukelen: A Neighborhood Epic  Film & Video   
1004298993  "Tamboura Plays Violin" - a collection of Pop ...         Music   
1004402863        Nightingale Noel - An A Cappella Holiday CD         Music   
1005185256                             Local Music Connection         Music   
100522240   DO NOT DUPLICATE - Selected by 2015 Devon Film...  Film & Video   
1005653464                                       SA-4 Studios         Music   
1006327667                           Letters to the Wild - EP         Music   
1008671527                          The Fitness Mindset   :-)    Publishing   
1008675685               Inspired & the Sleep EP - Eyelid Kid         Music   
1008815806  Into Winter: An Instrumental Christmas Experiment         Music   
1009154868       Help advance scientific guitar & bass design         Music   
1009312309                                        Mr. Pickles    Publishing   
1009774693                Universal Peace | Universal Justice         Music   
1010217550                               Solarboot-Havel-Tour         Music   
1010694326                                          Surrender         Music   
1011508551                               EBENEZER Solo Record         Music   
1011952280                         NHSOB Summer Festival 2015         Music   
1012058337                                        Degobah Max         Music   
1012414277    WE ARE WHO WE ARE BUT WHATEVER WE WANT TO BE!!!         Music   
1012704793                               Life In The Bus Lane  Film & Video   
1012744036          An Oratorio for our Time - Last Stop Cafe         Music   
1012807897                                  Juliet Remembered  Film & Video   
1013389538  Make admission to the London Music Awards cere...         Music   
...                                                       ...           ...   
982862687                                     Bydlo - A Novel    Publishing   
982919439                                               Faith  Film & Video   
983728691                    The Octopodes' 5th Studio Album!         Music   
985231591                                            TWO FISH  Film & Video   
986253191   DEVELOP MAGAZINE FOR THE ARTS, MUSIC, AND DESI...    Publishing   
986471034      Help Revival finish their new album, Guidance.         Music   
98744618                           Vinyl for Dancing Pigeons!         Music   
988763792     Maud Sings Maud: A Musical Betsy-Tacy Companion         Music   
989218116                    Keep on Movin' - The Debut Album         Music   
989499289                                           Sojourner         Music   
989874415                         Love and Soul through Sound         Music   
991409513                           Chris Copeland's First EP         Music   
991483230   Support Vocal Synergy's Second Semester and Te...         Music   
991820324                                   The Blend CD 2015         Music   
992433603                                   Jonny Come Lately  Film & Video   
994080962                                     Maximum Coppage  Film & Video   
994621918                                         Star Sailor  Film & Video   
994911312                     "Songwriter assistance" service         Music   
994933706                   

In [168]:
kickstarter_2018[kickstarter_2018.usd_pledged.isnull()].describe(include='all')

name category main_category currency  \
count                       3797     3797          3797     3797   
unique                      3764       28            12       10   
top     New EP/Music Development    Music         Music      USD   
freq                          28     2387          2388     2738   
first                        NaN      NaN           NaN      NaN   
last                         NaN      NaN           NaN      NaN   
mean                         NaN      NaN           NaN      NaN   
std                          NaN      NaN           NaN      NaN   
min                          NaN      NaN           NaN      NaN   
25%                          NaN      NaN           NaN      NaN   
50%                          NaN      NaN           NaN      NaN   
75%                          NaN      NaN           NaN      NaN   
max                          NaN      NaN           NaN      NaN   

                   deadline       goal             launched    pledged  \
count                  3797    3797.00                 3797    3797.00   
unique                  722        nan                 3797        nan   
top     2015-05-01 00:00:00        nan  2015-10-27 17:00:11        nan   
freq                     26        nan                    1        nan   
first   2012-04-24 00:00:00        nan  2012-03-14 19:05:44        nan   
last    2016-07-07 00:00:00        nan  2016-05-21 20:45:12        nan   
mean                    NaN    5413.35                  NaN    3023.26   
std                     NaN   37466.24                  NaN   44919.41   
min                     NaN     108.00                  NaN       0.00   
25%                     NaN    2000.00                  NaN       5.00   
50%                     NaN    3500.00                  NaN     640.00   
75%                     NaN    5000.00                  NaN    3845.00   
max                     NaN 1877000.00                  NaN 2762571.78   

            state  backers country  usd_pledged  usd_pledged_real  \
count        3797  3797.00    3797         0.00           3797.00   
unique          6      nan       1          nan               nan   
top     undefined      nan    N,0"          nan               nan   
freq         3562      nan    3797          nan               nan   
first         NaN      nan     NaN          nan               nan   
last          NaN      nan     NaN          nan               nan   
mean          NaN     0.00     NaN          nan           2949.85   
std           NaN     0.00     NaN          nan          44881.72   
min           NaN     0.00     NaN          nan              0.00   
25%           NaN     0.00     NaN          nan              5.00   
50%           NaN     0.00     NaN          nan            616.00   
75%           NaN     0.00     NaN          nan           3932.23   
max           NaN     0.00     NaN          nan        2762571.78   

        usd_goal_real  name_weirdness  
count         3797.00         3797.00  
unique            nan             nan  
top               nan             nan  
freq              nan             nan  
first             nan             nan  
last              nan             nan  
mean          5097.31           -0.01  
std          37258.79            0.12  
min             85.28           -3.00  
25%           2250.00            0.00  
50%           3583.05            0.00  
75%           5208.09            0.00  
max        1877000.00            0.00

Interesante, todos los datos pertenecen a un país indeterminado.

Todos los datos de ese "país" no tienen dinero prometido?

In [170]:
kickstarter_2018[(kickstarter_2018.country == 'N,0"') & ~(kickstarter_2018.usd_pledged.isnull())].head()

Empty DataFrame
Columns: [name, category, main_category, currency, deadline, goal, launched, pledged, state, backers, country, usd_pledged, usd_pledged_real, usd_goal_real, name_weirdness]
Index: []

A priori como los proyectos no tienen patrocinantes, completamos en 0 los valores

In [171]:
kickstarter_2018 = kickstarter_2018.fillna(0)
missing_values_count = kickstarter_2018.isnull().sum()
missing_values_count[missing_values_count > 0]

Series([], dtype: int64)

In [ ]:
kickstarter_2018.shape

Más métodos disponibles en https://pandas.pydata.org/pandas-docs/stable/api.html#api-dataframe-missing y http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

## 2.3. Codificar variables

Para trabajar con los algoritmos de aprendizaje automático, las variables categóricas estas deben ser codificadas como variables numéricas, no como cadenas.

Para esta tarea también hay diferentes estrategias, dos comunes son: asociar cadena a número y asociar cadena a columna.

In [ ]:
from sklearn import preprocessing

In [ ]:
kickstarter_2018.describe(include=['O'])

In [ ]:
column='main_category'
# Create a label (category) encoder object
le = preprocessing.LabelEncoder()
# Fit the encoder to the pandas column
le.fit(kickstarter_2018[column])

In [ ]:
# View encoder mapping
dict(zip(le.classes_,le.transform(le.classes_)))

In [ ]:
# Apply the fitted encoder to the pandas column
kickstarter_2018[column] = le.transform(kickstarter_2018[column]) 
kickstarter_2018.head()

In [ ]:
# Reversing encoding
kickstarter_2018[column] = le.inverse_transform(kickstarter_2018[column]) 
kickstarter_2018.head()

In [ ]:
from sklearn.preprocessing import LabelBinarizer
other_column = 'state'

lb = LabelBinarizer()

lb_results = lb.fit_transform(kickstarter_2018[other_column])
pd.DataFrame(lb_results, columns=((other_column + '_') + pd.Series(lb.classes_))).head(10)

More about preprocessing in http://scikit-learn.org/stable/modules/classes.html#module-sklearn.preprocessing

#### Ejercicio 4

Codificar `currency` con ambas estrategias

## 2.4. No cambiar los nombres de las variables de la fuente de origen


## 2.5. Verificar la consistencia de las variables
Aplicar reglas de integridad


In [ ]:
kickstarter_2018.head()

¿Hay proyectos sin patrocinadores pero con plata prometida?

In [ ]:
kickstarter_2018[(kickstarter_2018.backers == 0) & (kickstarter_2018.usd_pledged > 0)]

¿Hay proyecto no exitosos que pasaron el objetivo?

In [ ]:
kickstarter_2018[(kickstarter_2018.state != 'successful') & (kickstarter_2018.pledged > kickstarter_2018.goal)]

¿Cómo es, en promedio, la correlación entre el objetivo y la plata obtenida?

In [ ]:
df = kickstarter_2018.groupby('state')[['goal','pledged']].aggregate('mean')
df['pledged'] / df['goal']

¿Cuánta plata se obtiene en promedio por estado y categoría?

In [ ]:
kickstarter_2018.pivot_table(values='usd_pledged',index='state',columns='main_category')

#### Ejercicio 5

1. ¿Hay proyecto éxitosos que no consiguieron el objetivo? Si hay, ¿Qué porcentaje sí y cuál no?
2. Calcular una tabla con la cantidad de proyectos por categoría principal y estado.

## 2.6. Identificar y documentar valores atípicos/outliers


Queremos analizar las características de los proyectos que obtienen dinero pero sin considerar los casos atípicos.

In [ ]:
%matplotlib inline
kickstarter_2018.usd_pledged.hist();

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged');

"There are many rules of thumb for how to define an outlier but there is no silver bullet. One rule of thumb is that any data point that is three standard deviations away from the mean of the same data point for all observations."

In [ ]:
outliers = kickstarter_2018[kickstarter_2018.usd_pledged > (kickstarter_2018.usd_pledged.mean() + 
                                                            3 * kickstarter_2018.usd_pledged.std())]
outliers

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged');

In [ ]:
kickstarter_2018.drop(outliers.index).boxplot(column='usd_pledged',by='state');

Analizar outliers por categorías

In [ ]:
kickstarter_2018.boxplot(column='usd_pledged',by='state');
kickstarter_2018.boxplot(column='usd_goal_real',by='state');

#### Ejercicio 6

Calcular los valores atípicos de 'usd_goal_real' y graficar los boxplots, con y sin estos valores por categoría

## 2.7. Evaluar cómo comprimir los datos 
para su almacenamiento más eficiente.

En la actualidad, en las nubes el almacenamiento no suele ser problema sino la velocidad para leerlo que suele ser determinante a la hora de determinar el formato de almacenamiento.


## 2.8. Guardar el set de datos con un nombre informativo.


EL problema de la ingeniería de software